src/nupic/ntypes/Sdr.hpp

```diff
@@ -0,0 +1,707 @@
+        if( !dense_valid ) {
+            // Convert from flatIndex to dense.
+            dense.assign( size, 0 );
+            for(const auto idx : getFlatIndex()) {
+                dense[idx] = 1;
+            }
+            dense_valid_ = true;
+        }
+        return dense;
+    };
+
+    /**
+     * Gets the current value of the SDR.  The result of this method call is
+     * saved inside of this SDR until the SDRs value changes.  If the value is
+     * unset this raises an exception.
+     *
+     * @returns A pointer to a vector of the indices of the true values in the
+     * flattened SDR.
+     */
+    const vector<UInt>& getFlatIndex()
+        { return getFlatIndexMutable(); };
+
+    /**
+     * This method does the same thing as sdr.getFlatIndex() except that it
+     * returns a non-constant pointer.  After modifying the flatIndex vector you
+     * MUST call sdr.setFlatIndexInplace() in order to notify the SDR that its
+     * flatIndex vector has changed and its cached data is out of date.
+     *
+     * @returns A pointer to a vector of the indices of the true values in the
+     * flattened SDR.
+     */
+    vector<UInt>& getFlatIndexMutable() {
+        if( !flatIndex_valid ) {
+            flatIndex.clear(); // Clear out any old data.
+            if( index_valid ) {
+                // Convert from index to flatIndex.
+                const auto num_nz = size ? index[0].size() : 0;
+                flatIndex.reserve( num_nz );
+                for(UInt nz = 0; nz < num_nz; nz++) {
+                    UInt flat = 0;
+                    for(UInt dim = 0; dim < dimensions.size(); dim++) {
+                        flat *= dimensions[dim];
+                        flat += index[dim][nz];
+                    }
+                    flatIndex.push_back(flat);
+                }
+                flatIndex_valid_ = true;
+            }
+            else if( dense_valid ) {
+                // Convert from dense to flatIndex.
+                for(UInt idx = 0; idx < size; idx++)
+                    if( dense[idx] != 0 )
+                        flatIndex.push_back( idx );
+                flatIndex_valid_ = true;
+            }
+            else
+                throw logic_error("Can not get value from empty SDR.");
+        }
+        return flatIndex;
+    };
+
+    /**
+     * Gets the current value of the SDR.  The result of this method call is
+     * saved inside of this SDR until the SDRs value changes.  If the value is
+     * unset this raises an exception.
+     *
+     * @returns A pointer to a list of lists of the coordinates of the true
+     * values in the SDR.
+     */
+    const vector<vector<UInt>>& getIndex()
+        { return getIndexMutable(); };
+
+    /**
+     * This method does the same thing as sdr.getIndex() except that it
+     * returns a non-constant pointer.  After modifying the index vector you
+     * MUST call sdr.setIndexInplace() in order to notify the SDR that its
```

Comment:
Mutable views are needed because all of the setter methods copy the data. Mutable accessors are the only way to put data into the sdr without copying.

Reply:
I understand your point, but I'm still unhappy how this has to be used. 
Do you think this would be a better solution? 
- remove the `get*Mutable()` (keep `get{Sparse/Dense/..}()`) 
- make the set inplace methods take a vector that it just assigns (`setSparseInplace(&vector)`
That would make more clear what we do (set new vector)

Reply:
I fixed this by:
* Removed all mutable methods, all getter methods now return non-constant (mutable) references
* Overloaded setter methods to swap instead of copy where possible.  Swapping data vectors is very fast, but with the potential pitfall that it modifies both vectors.  It also requires that input vector be exactly the right data type.

The following is now valid code:
```
SDR X({99999});
auto & dense = X.getDense();
dense[2] = true;
X.setDense( dense );
```

Reply:
This is a really nice approach! Thanks for fixing that upto this state :clap: